File: two-sum.js
// Given an array of integers, return indices of the two numbers such that they add up to a specific target.
// You may assume that each input would have exactly one solution, and you may not use the same element twice.
// Example:
// Given nums = [2, 7, 11, 15], target = 9,
// Because nums[0] + nums[1] = 2 + 7 = 9,
// return [0, 1].


/* THOUGHTS */
// [2, 7, 11, 15], target = 9,

// result = {}, loop nums
// i = 0, currNum = 2, diff = target - currNum = 9 - 2 = 7

// if currNum is not in the result, add key value pair, diff: indexOf(currNum)
//  =>  result = {7: 0}

// i = 1, currNum = 7, diff = target - currNum = 9 - 7 = 2
// if currNum is in the result, return [indexOf(result[currNum]), i]

var twoSum = function(nums, target) {
  let result = new Map();
  
  for (let i in nums){
    let currNum = nums[i]
    if (result.has(currNum)) {
      return [result.get(currNum), i]
    } else {
      result.set((target-currNum), i)
    }
  }
}

File: TwoSum.py
# Brute Force approach O(N2) O(1)
# Loop through each element x and find another element that is target-x
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
      for i in range(len(nums)):
        for j in range(i+1, len(nums)):
          if nums[j] == target - nums[i]:
            return [i, j]


# HashMap approach O(N) O(N)
# loop nums, create a hashmap, key is target-nums[i], value is i
# hashmap = {
#     6-3: 0,
#     6-2: 1,
#     6-4: 2
# }

# if later num[i] is in keys, i and hashmap[num[i]] are the answer.
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dic = {}
        for i in range(len(nums)):
            diff = target - nums[i]
            if nums[i] in dic:
                return [dic[nums[i]], i]
            else:
                dic[diff] = i